### ISSUE TITLE
Was CQRS a good idea?

### ISSUE THREAD

**Issue by SebastianStehle** [Squidex/squidex#202]
This is a discussion about CQRS for Squidex. The question is: Is it a good idea and how do we have to change it. Some days ago I wanted to update the content UI: Lets say you only show the first 10 published content items and you unpublish an item. What should happen? I think you should just reload and get the updated 10 published items. But this is very hard with CQRS and eventual consistency. Because when you reload immediately the change might not have been propagated to the read model and you would get the same items again. Of course there are workaround: You can just wait a second or 2 and hope that it is propagated then. Or you can keep a list of unpublished ids in the UI. Then you load the first 20 items and filter out the just unpublished items. But what happens when the content has been published again by another user? Whatever you do: It can become complicated.

Let us have a look first why I have chosen this architecture:

1. **CQRS, Separation of write model and read model:** You can have a different model for your writes then for your reads. For example we use event sourcing for the write model and a document database for the read model.
2. **Event Sourcing, History and no model mismatch:** Event sourcing stored all events that ever happened in the system and derives the state from it. This might be slow, because there could be a lot of events. Therefore you typically use it with CQRS. Furthermore events are easy to serialize. You don’t have to think about how to store your business model in the database. 
3. **Eventual consistency:** Populating the read model might be slow and most systems do not support distributed transactions today. Therefore we populate the data in the background. This makes migration also very easy, because you just have to restart the processes to get fresh new databases. But in practice eventual consistency is problematic, because you are not used to it and we can not get all users to use it properly. For example when you have a simple blog application you would just redirect the user to the list of posts again when you have created a new blog post. The user would expect to see it immediately, but he probably has to refresh first.
4. **Events, Easy integration:** Events are a good place for integrations. Everything that happens in the system has an event and you can just listen to this events and make changes.

## Problems:

1. **Eventual consistency:** In practice eventual consistency is problematic, because you are not used to it and we can not get all users to use it properly. For example when you have a simple blog application you would just redirect the user to the list of posts again when you have created a new blog post. The user would expect to see it immediately, but he probably has to refresh first.
2. **Integration events:** In theory the events are a good place for integrations. Everything that happens in the system has an event and you can just listen to this events. In practice they are problematic, because they usually do not contain all the information. For example the ContentPublishedEvent only needs the content id (and some metadata). But when you want to forward it to another system you probably also want to have the username and the current data of the content. Then you have to pull for other information, which is slow very likely not possible because the data is not available yet.
3. **Scalability:** The background processes are hard to scale. At the moment there is one single threaded process for every target databases. But this also means that all updates to content items go through a single place. When one customer or user makes a bulk import other users get problems too. This can be solved, but is not that easy.

## What I would like to have

* Event Sourcing for history
* Immediate updates
* Easy migration process
* Better integration events

## How can we archive that? Some ideas:

1. Simple state persistence: For our domain objects we just use a single state persistent (aka snapshots) to store the state. This guarantees better consistency. Additionally we create events for background processes (e.g. statistics) or integrations (e.g. webhooks).
2. Eventual consistency for background processes: For derived data we can still use event sourcing. Such an example would be to populate a history view or send requests to webhooks.
3. Derive integration events: Another process creates integration events for the rule engine and pulls additional information, e.g. user names. We can just cache some of the data that does not change very often to achieve high throughput.
4. Migration tools: Migration of snapshots should be easy. You have to delete them and recreated them.

## Open Questions:

1. How can you write the events and the snapshot in the transaction or is it even necessary?

**Comment by jrgcubano** [comment]
Hi @SebastianStehle. Take a close look to https://github.com/eventflow/EventFlow to get some ideas!

**Comment by SebastianStehle** [comment]
Thank you, but it seems that they implement the "classic" CQRS model. Not sure what I can learn from it.

**Comment by dsbegnoche** [comment]
The EventFlow does implement Snapshots, so might be worth a look at how it is implemented there for some ideas. I would really like to see snapshots as it could really improve performance of playing events, and improve the system overall. 

An interesting article on the topic, it does not solve the problem, but has some interesting points. 
https://www.erikheemskerk.nl/event-sourcing-eventual-consistency-responding-to-events/

The only thing that has came to mind for me to resolve eventual consistency is to update the read store directly before returning from the controller.  This would make requests slower, and also not sure if it could lead to concurrency issues

**Comment by SebastianStehle** [comment]
I implemented Snapshots yesterday but mainly to simplify the persistence code.  But I am not sure, if it is really a big improvement: https://github.com/Squidex/squidex/blob/master/src/Squidex.Infrastructure/States/Persistence.cs

1. You only get benefits when making updates
2. You need a lot of events to see benefits (50, 100+).
3. You have to make 2 queries, which eats a lot of the performance improvements.

I think I read the aticle some month ago, don't know when exactly. But it makes my point clear. Deadling with eventual consistency is not easy. See my example above.

My first idea was to populate the read model directly. But as you said: This could lead to several problems because both writes (State and Events) are not transnational. Therefore you get inconsistencies.
My only solution so far is to write the events second and use it for less important stuff: https://github.com/Squidex/squidex/blob/refactoring-no-cqrs/src/Squidex.Infrastructure/Commands/DomainObjectBase.cs#L71 So it is a normal bug and you would deal with it the normal way, e.g. just write the content again or so.

When you write the events you have to deal with errors to the snapshot store and the question and then you would get other problems. e.g. Content looks unpublished and when you publish again you get a message: "Already published". You also happen at the moment. but there is a recovery mode. You can replay the events. The question is: What is the recovery mode for this scenario? How can you detect failures and write the snapshot again? (e.g. when the connection to MongoDB gets lost or for a bug).

**Comment by shiggan** [comment]
Another approach would be to put in infrastructure to allow the clients to solve this problem in a way that is more relevant for their problem domain

New tools
* Allow certain actions / schemas to have higher priorities than others (maybe a concept of a queue where actions/schemas get a faster response times – keeping in mind there will need to be some form of queue management).
* Publish event notifications over the web-service e.g. use signalr (or whatever the current hot thing is) to notify subscribers of relevant things that happened.
* Provide guidance around ways to accomplish the appearance of an immediate response (e.g. patterns around writing through a cache before writing to the squidex api; conversely my read actions would read through the cache before touching the squidex api) – there are plenty of ways to do this maybe document, or provide guidance around some options. Again this will be very specific to the use cases.

Use cases
1. My personal blog; I don’t care; I know the record will turn up in a few more seconds; basically what you do now is fine with me.
1. Site where certain types of content that I am writing should be prioritised over other writes – I would configure a ‘priority queue’ where creates/updates/deletes are prioritised over other actions.
In this case I would make use of your guidance around architecting around the eventual consistency problem.
1. Similar to 2; except rather than ‘faking’ an immediate user interface update I plug into a notification api in order to inform the user of a status of a change (success, in progress, error) – basically what the Azure portal does with its long running jobs.
1. A combination of 2 and 3; I have a queue configured in squidex; a write-through/read-through cache in place on my site; and I also subscribe to the notification api to help manage the state of my cache; and inform the user of useful things that happened.

To summarise: don't try and solve all the problems.

**Comment by SebastianStehle** [comment]
Thank you @neeasade  for your feedback.

I know that there are a lot of solutions and I also worked with SignalR in another CQRS project. A nice benefit was that you had a very good collaboration model, you could see all changes in your workspace almost instant.

The point is: It adds a lot of complexity and I am not sure if it is worth it. For example:

* I was also thinking about priority queues, but basically you need a queue or multiple queues per app.
* Therefore I would get a lot of queues and you have to manage them. You need a supervisor to create them dynamically. You also have to keep the number of subscriptions to the event store low (100, not 1000). 
* So you need some buffering and so on. Then it becomes really complicated.

For me, I see more disadvantages with the added complexity than benefits. And our time is limited, when we spend our time for infrastructure we will build a good CQRS framework but not tackle the features we would like to build.

**Comment by pushrbx** [comment]
I throw my idea in the hat too for the eventual consistency problem: **The ignore way** - I think this problem can be solved with a basic tool which shows for the user/customer of squidex how to implement a client for it: I can imagine a basic angular library which would have the required functionality to indicate that the posted data will be there eventually just wait for it. Also don't forget, that the main user base that squidex tries to target is mostly consist of software developers, and you can safely assume that they will make the right judgement to use squidex or not, seeing that it works this way.

I agree with @shiggan, don't try to solve all these problems unless it is really necessary. Can we live with them easily? If yes, then leave it.

**Comment by SebastianStehle** [comment]
I cannot ignore it, because I also have problems with it: Scalability and implementation details.

**Comment by mjhpour** [comment]
What if we use Reactive Extensions inside CQRS?!

**Comment by SebastianStehle** [comment]
How would that Help?
